File: TCPclient.c
/*
 * TCPclient.c
 * Systems and Networks II
 * Project 1
 *
 * This file describes the functions to be implemented by the TCPclient.
 * You may also implement any auxillary functions you deem necessary.
 */
 
#include <stdio.h> 			/* for printf() and fprintf() */
#include <sys/socket.h>			/* for socket(), connect(), send(), and recv() */
#include <arpa/inet.h>			/* for sockaddr_in and inet_addr() */
#include <stdlib.h>			/* for atoi() and exit() */
#include <string.h>			/* for memset() */
#include <unistd.h>			/* for close() */
#include "TCPclient.h"

#define RCVBUFSIZE 256   /* Size of receive buffer */

/*
 * Creates a streaming socket and connects to a server.
 *
 * serverName - the ip address or hostname of the server given as a string
 * port       - the port number of the server
 * dest       - the server's address information; the structure should be created with information
 *              on the server (like port, address, and family) in this function call
 *
 * return value - the socket identifier or a negative number indicating the error if a connection could not be established
 */
int createSocket(char * serverName, int port, struct sockaddr_in * dest)
{
		int socketCreation = 0;
		memset(dest, 0, sizeof(dest));
		dest.sin_family = AF_INET;
		inet_pton(AF_INET, serverName, dest.sin_addr); //Sever IP address
		dest.sin_port = hton(port);
		socketCreation = socket(PF_INET, SOCK_STREAM, 0);
		if(connect(s, (struct sockaddr *)dest, sizeof(dest))< 0)
		{
			socketCreation = -1;
		}
		return socketCreation;
}

/*
 * Sends a request for service to the server. This is an asynchronous call to the server, 
 * so do not wait for a reply in this function.
 * 
 * sock    - the socket identifier
 * request - the request to be sent encoded as a string
 * dest    - the server's address information
 *
 * return   - 0, if no error; otherwise, a negative number indicating the error
 */
int sendRequest(int sock, char * request, struct sockaddr_in * dest)
{
	int error = 0;
	error = write(sock, request, strlen(request));
	return error;
}	

/*
 * Receives the server's response formatted as an XML text string.
 *
 * sock     - the socket identifier
 * response - the server's response as an XML formatted string to be filled in by this function;
 *            memory is allocated for storing response
 *
 * return   - 0, if no error; otherwise, a negative number indicating the error
 */
int receiveResponse(int sock, char * response)
{
	int error = 0;
	error = read(sock, response, RCVBUFSIZE);
	return error;
}

/*
 * Prints the response to the screen in a formatted way.
 *
 * response - the server's response as an XML formatted string
 *
 */
void printResponse(char* response)
{
		printf("\nResponse is: %s\n", response);

		char * temp;
		temp = strtok(response, " .\n");
		while(temp != NULL)
		{
			
			if((strstr(temp, "<reply>") == NULL) && (strstr(temp, "</reply>") == NULL))
			{
				printf("%s ", temp);
			}
			temp = strtok('\0', " .\n");
		}
		printf("\n\n");
}

/*
 * Closes the specified socket
 *
 * sock - the ID of the socket to be closed
 * 
 * return - 0, if no error; otherwise, a negative number indicating the error
 */
int closeSocket(int sock)
{
	int error = 0;
	error = close(sock);
	return error;
}

int main(int argc, char ** argv)
{
	int sock;                        /* Socket descriptor */
    struct sockaddr_in * ServAddr; 	 /* Server address */
    unsigned short echoServPort;     /* Echo server port */
    char * servIP;                   /* Server IP address (dotted quad) */
    char * stringSent;               /* String to send to server */
    char * stringReceived			 /* String send back to client*/
    char echoBuffer[RCVBUFSIZE];     /* Buffer for echo string */
    unsigned int echoStringLen;      /* Length of string to echo */
    int bytesRcvd, totalBytesRcvd;   /* Bytes read in single recv() 
                                        and total bytes read */
    int portNum, temp;

    stringSent = (char *)malloc(1+150*(sizeof(char)));
    stringReceived = (char *)malloc(1+150*(sizeof(char)));
    strcpy(stringSent, "<reply> Here is a string! </reply>\n");
	printf("String sent: %s\n\n", stringSent);
	printResponse(temp);

	 if(argc != 2)
	{
		perror("There is not enough arguments in entered in\n");
		exit(1);
	}
	servIP = argv[0];
	portNum = atoi(argv[1]);
	sock = createSocket(servIP, portNum, ServAddr);
	if(sock < 0)
	{
		perror("Could not create socket\n");
		exit(1);
	}
	temp = sendRequest(sock, stringSent, ServAddr);
	if(temp != 0)
	{
		perror("Could not send request\n");
		exit(1);
	}
	temp = receiveResponse(sock, stringReceived);
	if(temp != 0)
	{
		perror("Could not receive string\n");
		exit(1);
	}
	printResponse(stringReceived);
	
	free(stringSent);
	free(stringReceived);
	
	return 0;
}

void emptyBuffer(char * arry, int sizeArry)
{
	int i;
	for(i = 0; i < sizeArry; i++)
	{
		arry[i] = '\0';	
	}	
}

void showASCII(char * str)
{
    int i, sentSize, j;
    sentSize = strlen(str);
    sentSize = sentSize + 3;
    for(i = 0; i < sentSize; i++)
    {
      	printf("%d ", str[i]);   
    }
    printf("\n");
}

File: TCPclient.java
/*
 * TCPclient.java
 * Systems and Networks II
 * Project 1
 *
 * This file describes the functions to be implemented by the TCPclient class
 * You may also implement any auxiliary functions you deem necessary.
 */
import java.io.*;
import java.net.*;



public class TCPclient
{
	private Socket _socket; // the socket for communication with a server
	private static String sendString; //the string sent to the server
	private static String receivedString; //the string received from the server
	private BufferedReader stringIn;
	private BufferedWriter stringOut;
	private PrintWriter out;
	
	/**
	 * Constructs a TCPclient object.
	 */
	public TCPclient()
	{
		sendString = new String("<reply> hello world </reply>");
		receivedString = new String();
	}
	
	/**
	 * Creates a streaming socket and connects to a server.
	 *
	 * @param host - the ip or hostname of the server
	 * @param port - the port number of the server
	 *
	 * @return - 0 or a negative number describing an error code if the connection could not be established
	 */
	public int createSocket(String host, int port)
	{
		int error = 0;
		try {
			_socket = new Socket(host, port);
		} catch (IOException e) {
			error = -1;
			e.printStackTrace();
		}
		return error;
	}

	/**
	 * Sends a request for service to the server. Do not wait for a reply in this function. This will be
	 * an asynchronous call to the server.
	 * 
	 * @request - the request to be sent
	 *
	 * @return - 0, if no error; otherwise, a negative number indicating the error
	 */
	public int sendRequest(String request)
	{
		int error = 0;
		try {
			out = new PrintWriter(_socket.getOutputStream(), true);
			out.println(request);
		} catch (IOException e) {
			System.err.println("Error could not send request to server\n");
			e.printStackTrace();
		}
		return error;
	}
	
	/**
	 * Receives the server's response. Also displays the response in a
	 * clear and concise format.
	 *
	 * @return - the server's response or NULL if an error occured
	 */
	public String receiveResponse()
	{
		String response = new String(); 
		try {
			stringIn = new BufferedReader(new InputStreamReader(_socket.getInputStream()));
		} catch (IOException e) {
			System.err.println("Error: could not receive input from server\n");
			e.printStackTrace();
		}
		response = stringIn.toString();
		return response;
	}
	
	/*
    * Prints the response to the screen in a formatted way.
    *
    * response - the server's response as an XML formatted string
    *
    */
	public static void printResponse(String response)
	{
		System.out.println(response + "\n");
	}
 

	/*
	 * Closes an open socket.
	 *
    * @return - 0, if no error; otherwise, a negative number indicating the error
	 */
	public int closeSocket() 
	{
		int error = 0;
		try {
			_socket.close();
		} catch (IOException e) {
			error = -1;
			e.printStackTrace();
		}
		return error;
	}

	/**
	 * The main function. Use this function for 
	 * testing your code. We will use our own main function for testing.
	 */
	public static void main(String[] args)
	{
		if(args.length != 2)
		{
			System.err.println("Error: Incorrect number of arguments must be <hostname> <portnumber>");
			System.exit(1);
		}
		String host = args[0];
		int port = Integer.parseInt(args[2]); 
		TCPclient client = new TCPclient();
		client.createSocket(host, port);
		client.sendRequest(sendString);
		receivedString = client.receiveResponse();
		client.closeSocket();
	
	}

}
